benchmarks/Kestrel.Performance/InMemoryTransportBenchmark.cs

```diff
@@ -0,0 +1,196 @@
+            await _connection.SendRequestAsync(RequestParsingData.PlaintextTechEmpowerPipelinedRequests);
+            await _connection.GetResponseAsync(_plaintextPipelinedExpectedResponseLength);
+        }
+
+        public class InMemoryTransportFactory : ITransportFactory
+        {
+            private readonly int _connectionsPerEndPoint;
+
+            private readonly Dictionary<IEndPointInformation, IReadOnlyList<InMemoryConnection>> _connections =
+                new Dictionary<IEndPointInformation, IReadOnlyList<InMemoryConnection>>();
+
+            public IReadOnlyDictionary<IEndPointInformation, IReadOnlyList<InMemoryConnection>> Connections => _connections;
+
+            public InMemoryTransportFactory(int connectionsPerEndPoint)
+            {
+                _connectionsPerEndPoint = connectionsPerEndPoint;
+            }
+
+            public ITransport Create(IEndPointInformation endPointInformation, IConnectionHandler handler)
+            {
+                var connections = new InMemoryConnection[_connectionsPerEndPoint];
+                for (var i = 0; i < _connectionsPerEndPoint; i++)
+                {
+                    connections[i] = new InMemoryConnection();
+                }
+
+                _connections.Add(endPointInformation, connections);
+
+                return new InMemoryTransport(handler, connections);
+            }
+        }
+
+        public class InMemoryTransport : ITransport
+        {
+            private readonly IConnectionHandler _handler;
+            private readonly IReadOnlyList<InMemoryConnection> _connections;
+
+            public InMemoryTransport(IConnectionHandler handler, IReadOnlyList<InMemoryConnection> connections)
+            {
+                _handler = handler;
+                _connections = connections;
+            }
+
+            public Task BindAsync()
+            {
+                foreach (var connection in _connections)
+                {
+                    _handler.OnConnection(connection);
+                }
+
+                return Task.CompletedTask;
+            }
+
+            public Task StopAsync()
+            {
+                return Task.CompletedTask;
+            }
+
+            public Task UnbindAsync()
+            {
+                return Task.CompletedTask;
+            }
+        }
+
+        public class InMemoryConnection : TransportConnection
+        {
+            public Task SendRequestAsync(byte[] request)
+            {
+                return Input.WriteAsync(request);
+            }
+
+            public async Task<byte[]> GetResponseAsync(int length)
+            {
+                while (true)
+                {
+                    var result = await Output.ReadAsync();
+                    var buffer = result.Buffer;
+                    var consumed = buffer.Start;
+                    var examined = buffer.End;
+
+                    try
+                    {
+                        if (buffer.Length >= length)
```

Comment:
Let's change this to consume (AdvanceTo end) every time, it reflects better what socket/libuv layer is doing.

Reply:
Can you specify exactly what code should be changed to what other code?

Reply:
Roughly:

```
while (length > 0)
{
var result = await Output.ReadAsync();
var buffer = result.Buffer;
length -= buffer.Lenght;
Output.AdvanceTo(buffer.End}
}

if (length< 0 ) throw InvalidOperationException("Unexpected responce");
```